File: src/lib/api/calculator.ts
import type { ParametriFinanziari, Valutazione, CashflowProjection, BusinessPlanResult, IndipendenzaFinanziaria } from "@/types";

export class CalcolatoreFinanziario {
  private aliquota_contributo: number;
  private aliquota_finanziamento: number;
  private limite_min: number;
  private limite_max: number;
  private fatturato_minimo: number;
  private bilanci_richiesti: number;

  constructor(parametri: ParametriFinanziari) {
    this.aliquota_contributo = parametri.aliquota_contributo || 0;
    this.aliquota_finanziamento = parametri.aliquota_finanziamento || 0;
    this.limite_min = parametri.limite_min_investimento || 0;
    this.limite_max = parametri.limite_max_investimento || 0;
    this.fatturato_minimo = parametri.fatturato_minimo || 0;
    this.bilanci_richiesti = parametri.bilanci_richiesti || 0;
  }

  calcola(investimento: number) {
    if (this.limite_min > 0 && investimento < this.limite_min) {
      return {
        successo: false,
        errore: `Investimento (€${investimento.toLocaleString("it-IT")}) < minimo (€${this.limite_min.toLocaleString("it-IT")})`,
        contributo: 0, finanziamento: 0, totale_agevolabile: 0,
      };
    }
    let troncato = false;
    if (this.limite_max > 0 && investimento > this.limite_max) {
      investimento = this.limite_max;
      troncato = true;
    }
    const contributo = investimento * (this.aliquota_contributo / 100);
    const finanziamento = investimento * (this.aliquota_finanziamento / 100);
    return {
      successo: true,
      troncato,
      investimento_effettivo: investimento,
      contributo,
      finanziamento,
      totale_agevolabile: contributo + finanziamento,
      aliquota_contributo: this.aliquota_contributo,
      aliquota_finanziamento: this.aliquota_finanziamento,
    };
  }

  valida_bilanci(data_costituzione: string, bilanci_depositati: number): Valutazione {
    if (this.bilanci_richiesti <= 0)
      return { conforme: true, stato: "N/D", dettaglio: "Nessun vincolo bilanci" };
    if (bilanci_depositati < this.bilanci_richiesti)
      return { conforme: false, stato: "ROSSO", dettaglio: `Servono ${this.bilanci_richiesti} bilanci, disponibili ${bilanci_depositati}` };
    return { conforme: true, stato: "VERDE", dettaglio: `${bilanci_depositati}/${this.bilanci_richiesti} bilanci OK` };
  }

  valida_fatturato(fatturato: number): Valutazione {
    if (this.fatturato_minimo <= 0)
      return { conforme: true, stato: "N/D", dettaglio: "Nessun vincolo fatturato" };
    if (fatturato < this.fatturato_minimo)
      return { conforme: false, stato: "ROSSO", dettaglio: `Fatturato €${fatturato.toLocaleString("it-IT")} < minimo €${this.fatturato_minimo.toLocaleString("it-IT")}` };
    if (fatturato < this.fatturato_minimo * 1.2)
      return { conforme: true, stato: "GIALLO", dettaglio: `Margine sicurezza insufficiente: €${fatturato.toLocaleString("it-IT")} vs €${this.fatturato_minimo.toLocaleString("it-IT")}` };
    return { conforme: true, stato: "VERDE", dettaglio: "Fatturato conforme" };
  }
}

export function calcolaBusinessPlan(
  investimento: number,
  contributo: number,
  finanziamento: number,
  utileNetto?: number,
): BusinessPlanResult {
  const anni = 5;
  const tassoSconto = 0.08;
  const rataFinanziamento = anni > 0 ? finanziamento / anni : 0;

  const cashflow: CashflowProjection[] = [];
  let cumulato = -investimento + contributo;
  let paybackAnni = -1;

  for (let anno = 1; anno <= anni; anno++) {
    const crescita = 0.15 + (anno - 1) * 0.05;
    const ricavi = investimento * crescita;
    const costiOp = investimento * 0.08;
    const quotaFin = rataFinanziamento;
    const ammortamento = investimento / anni;
    const costi = costiOp + quotaFin + ammortamento;
    const netto = ricavi - costi;
    cumulato += netto;
    if (paybackAnni < 0 && cumulato >= 0) paybackAnni = anno;
    cashflow.push({
      anno,
      ricavi: Math.round(ricavi),
      costi: Math.round(costi),
      netto: Math.round(netto),
    });
  }

  if (paybackAnni < 0) paybackAnni = anni + 1;

  // DSCR: utile_netto / rata_finanziamento (fallback a EBITDA medio se utile_netto non disponibile)
  const ebitdaMedio = cashflow.reduce((acc, c) => acc + (c.ricavi - Math.round(investimento * 0.08)), 0) / anni;
  const dscrBase = (utileNetto != null && utileNetto > 0) ? utileNetto : ebitdaMedio;
  const dscr = rataFinanziamento > 0 ? dscrBase / rataFinanziamento : 0;

  const flussi = cashflow.map((c) => c.netto);
  const van = -investimento + flussi.reduce((acc, val, i) => acc + val / Math.pow(1 + tassoSconto, i + 1), 0);
  const irr = calcolaIrrSicuro([-investimento, ...flussi]);

  return {
    dscr: Math.round(dscr * 100) / 100,
    payback_anni: paybackAnni,
    van: Math.round(van),
    irr: Math.round(irr * 100) / 100,
    cashflow,
    contributo: Math.round(contributo),
    finanziamento: Math.round(finanziamento),
    investimento_totale: Math.round(investimento),
  };
}

export function calcolaPayback(
  investimento: number,
  contributo: number,
  rataFinanziamento: number,
  proiezioni: { netto: number }[],
): { anni: number; raggiunto: boolean; messaggio: string } {
  let cumulato = -investimento + contributo;
  const anni = proiezioni.length;
  for (let i = 0; i < anni; i++) {
    cumulato += proiezioni[i].netto;
    if (cumulato >= 0) {
      return { anni: i + 1, raggiunto: true, messaggio: `Payback a ${i + 1} anni` };
    }
  }
  return { anni: anni + 1, raggiunto: false, messaggio: `Payback > ${anni} anni — non raggiunto nel periodo` };
}

function npv(rate: number, flussi: number[]): number {
  return flussi.reduce((acc, val, i) => acc + val / Math.pow(1 + rate, i), 0);
}

function calcolaIrrSicuro(flussi: number[]): number {
  const precision = 1e-6;
  const maxIter = 200;

  // Newton
  let x = 0.1;
  for (let iter = 0; iter < maxIter; iter++) {
    const f1 = npv(x, flussi);
    const f2 = flussi.reduce((acc, val, i) => acc - (i * val) / Math.pow(1 + x, i + 1), 0);
    if (Math.abs(f2) < 1e-12) break;
    const xNew = x - f1 / f2;
    if (Math.abs(xNew - x) < precision) {
      if (xNew > -0.999 && xNew < 10) return xNew;
      break;
    }
    x = xNew;
  }

  // Fallback bisezione
  let lo = -0.99, hi = 10.0;
  let fLo = npv(lo, flussi);
  let fHi = npv(hi, flussi);
  if (fLo * fHi > 0) return 0;
  for (let i = 0; i < 100; i++) {
    const mid = (lo + hi) / 2;
    const fMid = npv(mid, flussi);
    if (Math.abs(fMid) < precision) return mid;
    if (fLo * fMid <= 0) { hi = mid; fHi = fMid; }
    else { lo = mid; fLo = fMid; }
  }
  return 0;
}

export function calcolaIndipendenzaFinanziaria(
  patrimonioNetto: number,
  debitiFinanziari: number,
): IndipendenzaFinanziaria {
  if (!debitiFinanziari || debitiFinanziari <= 0) {
    return { indice: 1.0, stato: "VERDE", dettaglio: "Nessun debito finanziario" };
  }
  const indice = patrimonioNetto / debitiFinanziari;
  if (indice >= 1.0) {
    return { indice: Math.round(indice * 100) / 100, stato: "VERDE", dettaglio: `Patrimonio netto copre i debiti finanziari (x${indice.toFixed(2)})` };
  }
  if (indice >= 0.5) {
    return { indice: Math.round(indice * 100) / 100, stato: "GIALLO", dettaglio: `Patrimonio netto copre il ${Math.round(indice * 100)}% dei debiti` };
  }
  return { indice: Math.round(indice * 100) / 100, stato: "ROSSO", dettaglio: `Patrimonio netto insufficiente — copre solo il ${Math.round(indice * 100)}% dei debiti` };
}

File: backend/services/calculator.py
from datetime import datetime


class CalcolatoreFinanziario:
    def __init__(self, parametri: dict):
        self.aliquota_contributo = parametri.get("aliquota_contributo", 0) or 0
        self.aliquota_finanziamento = parametri.get("aliquota_finanziamento", 0) or 0
        self.limite_min = parametri.get("limite_min_investimento", 0) or 0
        self.limite_max = parametri.get("limite_max_investimento", 0) or 0
        self.fatturato_minimo = parametri.get("fatturato_minimo", 0) or 0
        self.bilanci_richiesti = parametri.get("bilanci_richiesti", 0) or 0

    def calcola(self, investimento: float) -> dict:
        if self.limite_min > 0 and investimento < self.limite_min:
            return {
                "successo": False,
                "errore": f"Investimento ({investimento:,.0f}€) inferiore al minimo ({self.limite_min:,.0f}€)",
                "contributo": 0,
                "finanziamento": 0,
                "totale_agevolabile": 0,
            }
        troncato = False
        if self.limite_max > 0 and investimento > self.limite_max:
            investimento = float(self.limite_max)
            troncato = True
        contributo = investimento * (self.aliquota_contributo / 100)
        finanziamento = investimento * (self.aliquota_finanziamento / 100)
        return {
            "successo": True,
            "troncato": troncato,
            "investimento_effettivo": round(investimento, 2),
            "contributo": round(contributo, 2),
            "finanziamento": round(finanziamento, 2),
            "totale_agevolabile": round(contributo + finanziamento, 2),
            "aliquota_contributo": self.aliquota_contributo,
            "aliquota_finanziamento": self.aliquota_finanziamento,
        }

    def valida_bilanci(self, data_costituzione: str, anni_bilancio: int) -> dict:
        if self.bilanci_richiesti <= 0:
            return {"conforme": True, "stato": "N/D", "dettaglio": "Nessun vincolo"}
        if anni_bilancio < self.bilanci_richiesti:
            return {"conforme": False, "stato": "ROSSO",
                    "dettaglio": f"Servono {self.bilanci_richiesti} bilanci, disponibili {anni_bilancio}"}
        return {"conforme": True, "stato": "VERDE",
                "dettaglio": f"{anni_bilancio}/{self.bilanci_richiesti} bilanci OK"}

    def valida_fatturato(self, fatturato: float) -> dict:
        if self.fatturato_minimo <= 0:
            return {"conforme": True, "stato": "N/D", "dettaglio": "Nessun vincolo"}
        if fatturato < self.fatturato_minimo:
            return {"conforme": False, "stato": "ROSSO",
                    "dettaglio": f"Fatturato {fatturato:,.0f}€ < minimo {self.fatturato_minimo:,.0f}€"}
        if fatturato < self.fatturato_minimo * 1.2:
            return {"conforme": True, "stato": "GIALLO",
                    "dettaglio": f"Margine insufficiente: {fatturato:,.0f}€ vs {self.fatturato_minimo:,.0f}€"}
        return {"conforme": True, "stato": "VERDE", "dettaglio": "Fatturato conforme"}


def calcola_business_plan(
    investimento: float,
    contributo: float,
    finanziamento: float,
    utile_netto: float = 0,
) -> dict:
    try:
        from numpy_financial import irr as np_irr
        USE_NUMPY = True
    except ImportError:
        USE_NUMPY = False

    anni = 5
    tasso_sconto = 0.08
    rata_finanziamento = finanziamento / anni if anni > 0 else 0
    cashflow = []
    cumulato = -investimento + contributo
    payback_anni = -1

    for anno in range(1, anni + 1):
        crescita = 0.15 + (anno - 1) * 0.05
        ricavi = investimento * crescita
        costi_op = investimento * 0.08
        quota_fin = rata_finanziamento
        ammortamento = investimento / anni
        costi = costi_op + quota_fin + ammortamento
        netto = ricavi - costi
        cumulato += netto
        if payback_anni < 0 and cumulato >= 0:
            payback_anni = anno
        cashflow.append({
            "anno": anno,
            "ricavi": round(ricavi),
            "costi": round(costi),
            "netto": round(netto),
        })

    if payback_anni < 0:
        payback_anni = anni + 1

    # DSCR: utile_netto / rata_finanziamento (fallback a EBITDA medio se utile_netto non disponibile)
    ebitda_medio = sum(c["ricavi"] - investimento * 0.08 for c in cashflow) / anni
    dscr_base = utile_netto if utile_netto > 0 else ebitda_medio
    dscr = dscr_base / rata_finanziamento if rata_finanziamento > 0 else 0

    flussi_netti = [c["netto"] for c in cashflow]
    flussi_irr = [-investimento] + flussi_netti
    van = -investimento + sum(v / (1 + tasso_sconto) ** (i + 1) for i, v in enumerate(flussi_netti))

    if USE_NUMPY and len(flussi_irr) >= 2:
        try:
            raw = float(np_irr(flussi_irr))
            irr = max(-0.999, min(raw, 10.0))
        except Exception:
            irr = _calcola_irr_sicura(flussi_irr)
    else:
        irr = _calcola_irr_sicura(flussi_irr)

    return {
        "dscr": round(dscr, 2),
        "payback_anni": payback_anni,
        "van": round(van),
        "irr": round(irr * 100, 2),
        "cashflow": cashflow,
        "contributo": round(contributo),
        "finanziamento": round(finanziamento),
        "investimento_totale": round(investimento),
    }


def _calcola_irr_sicura(flussi: list) -> float:
    """IRR via Newton con protezione convergenza, fallback a bisezione."""
    precision = 1e-6
    max_iter = 200

    def npv(rate):
        return sum(v / (1 + rate) ** i for i, v in enumerate(flussi))

    def npv_deriv(rate):
        return sum(-i * v / (1 + rate) ** (i + 1) for i, v in enumerate(flussi))

    # Newton
    x = 0.1
    for _ in range(max_iter):
        f = npv(x)
        df = npv_deriv(x)
        if abs(df) < 1e-12:
            break
        x_new = x - f / df
        if abs(x_new - x) < precision:
            if -0.999 < x_new < 10.0:
                return x_new
            break
        x = x_new

    # Fallback: bisezione
    lo, hi = -0.99, 10.0
    f_lo = npv(lo)
    f_hi = npv(hi)
    if f_lo * f_hi > 0:
        return 0.0
    for _ in range(100):
        mid = (lo + hi) / 2
        f_mid = npv(mid)
        if abs(f_mid) < precision:
            return mid
        if f_lo * f_mid <= 0:
            hi = mid
            f_hi = f_mid
        else:
            lo = mid
            f_lo = f_mid
    return 0.0


def calcola_payback(
    investimento: float,
    contributo: float,
    rata_finanziamento: float,
    proiezioni: list,
) -> dict:
    """Calcola il payback period in anni basato sulle proiezioni di cashflow."""
    cumulato = -investimento + contributo
    anni = len(proiezioni)
    for i, p in enumerate(proiezioni):
        netto = p.get("netto", 0) if isinstance(p, dict) else p
        cumulato += netto
        if cumulato >= 0:
            return {"anni": i + 1, "raggiunto": True, "messaggio": f"Payback a {i + 1} anni"}
    return {"anni": anni + 1, "raggiunto": False, "messaggio": f"Payback > {anni} anni — non raggiunto nel periodo"}


def calcola_indipendenza_finanziaria(patrimonio_netto: float, debiti_finanziari: float) -> dict:
    if debiti_finanziari <= 0:
        return {"indice": 1.0, "stato": "VERDE", "dettaglio": "Nessun debito finanziario"}
    indice = patrimonio_netto / debiti_finanziari if debiti_finanziari > 0 else 0
    if indice >= 1.0:
        stato = "VERDE"
        dettaglio = f"Indice {indice:.2f} — patrimonio netto copre i debiti finanziari"
    elif indice >= 0.5:
        stato = "GIALLO"
        dettaglio = f"Indice {indice:.2f} — patrimonio netto copre il 50%+ dei debiti"
    else:
        stato = "ROSSO"
        dettaglio = f"Indice {indice:.2f} — patrimonio netto insufficiente a coprire i debiti"
    return {"indice": round(indice, 2), "stato": stato, "dettaglio": dettaglio}
